### PR TITLE
[SU-108] Add endpoint to unlink NIH account

### PR DESCRIPTION
**Pull request by MatthewBemis**
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/resources/swagger/api-docs.yaml [hunk 1]
```diff
@@ -3133,6 +3133,22 @@ paths:
           description: Internal server error retrieving NIH status
           content: {}
       x-passthrough: false
+  /api/nih/unlink:
+    delete:
```
Comment (nawatts) [inline src/main/resources/swagger/api-docs.yaml:3137]: Nit: `delete unlink` seems like a double negative. Would `delete link` or `post unlink` be a bit more conventional?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala [hunk 2]
```diff
@@ -127,6 +128,29 @@ class NihService(val samDao: SamDAO, val thurloeDao: ThurloeDAO, val googleDao:
     thurloeDao.saveKeyValues(userInfo, profilePropertyMap)
   }
 
+  private def unlinkNihAccount(userInfo: UserInfo): Future[Unit] = {
+    val nihKeys = Set("linkedNihUsername", "linkExpireTime")
+
+    Future.traverse(nihKeys) { nihKey =>
+      thurloeDao.deleteKeyValue(userInfo.id, nihKey, userInfo)
+    } map { results =>
+      if(!results.forall(_.isSuccess))
```
Comment (nawatts) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala]: Is it a concern if only one key is successfully deleted? Does Thurloe provide an option to delete multiple keys in the same request?

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpThurloeDAO.scala [hunk 3]
```diff
@@ -80,6 +80,15 @@ class HttpThurloeDAO ( implicit val system: ActorSystem, implicit val executionC
     saveKeyValues(userInfo, profilePropertyMap).map(_ => ())
   }
 
+  override def deleteKeyValue(forUserId: String, keyName: String, callerToken: WithAccessToken): Future[Try[Unit]] = {
```
Comment (nawatts) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpThurloeDAO.scala:83]: Out of curiosity, why is the return value here wrapped in a `Try`?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala [hunk 4]
```diff
@@ -127,6 +128,29 @@ class NihService(val samDao: SamDAO, val thurloeDao: ThurloeDAO, val googleDao:
...
+        throw new FireCloudExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, "Unable to unlink NIH account"))
+    }
+  }
+
+  def unlinkNihAccountAndSyncSelf(userInfo: UserInfo): Future[Unit] = {
+    for {
+      _ <- unlinkNihAccount(userInfo)
+      _ <- ensureWhitelistGroupsExists()
+      _ <- Future.traverse(nihWhitelists) {
+        whitelist => removeUserFromNihWhitelistGroup(WorkbenchEmail(userInfo.userEmail), whitelist).recoverWith {
```
Comment (nawatts) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala:152]: Similar to above, is it a concern if deleting the keys in Thurloe succeeds but removing from the groups fails?

**Review by nawatts — APPROVED** [review]

File: src/main/resources/swagger/api-docs.yaml (hunk 1 above)
Comment (MatthewBemis) [reply src/main/resources/swagger/api-docs.yaml:3137]: I agree, I don't like putting the verb in the path either. I mainly did this to keep it consistent with how other endpoints are structured throughout orch/rawls. I'd be happier if it was `DELETE /api/nih/account`  which is more RESTful.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala (hunk 2 above)
Comment (MatthewBemis) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala]: I might be able to do this with https://thurloe.dsde-dev.broadinstitute.org/#/Key-value%20pairs/setKey

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala (hunk 4 above)
Comment (MatthewBemis) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala:152]: Deleting the keys but not removing from the group is okay imo, the process that syncs NIH groups will realize it within 6 hours and correct it.

File: src/main/resources/swagger/api-docs.yaml (hunk 1 above)
Comment (MatthewBemis) [reply src/main/resources/swagger/api-docs.yaml:3137]: Made it nice in 6c7c6a0858d7a849bccd80daa7960d679fadd365

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala (hunk 2 above)
Comment (MatthewBemis) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala]: That endpoint doesn't appear to do what I want (it doesn't support removing keys), so we're stuck with the 2 calls to deleteKey. It isn't a concern though- we will still be able to properly update access to the NIH groups if only one key is removed. The only side effect will be that the user will see a value for the un-deleted key in their profile, but they can unlink again to clear it if they'd like.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpThurloeDAO.scala (hunk 3 above)
Comment (MatthewBemis) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpThurloeDAO.scala:83]: Solely for consistency with the other DAO method that manipulates keys

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala [hunk 5]
```diff
@@ -127,6 +128,29 @@
...
   }
 
+  private def unlinkNihAccount(userInfo: UserInfo): Future[Unit] = {
+    val nihKeys = Set("linkedNihUsername", "linkExpireTime")
+
+    Future.traverse(nihKeys) { nihKey =>
+      thurloeDao.deleteKeyValue(userInfo.id, nihKey, userInfo)
+    } map { results =>
+      if(!results.forall(_.isSuccess))
+        throw new FireCloudExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, "Unable to unlink NIH account"))
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala]: It looks like this is swallowing the exception thrown by `ThurloeDao.deleteKeyValue`, and will lose the key that failed to be deleted.

**Review by tlangs — CHANGES_REQUESTED** [review]
Looks good, but I'm worried about losing information in the exceptions thrown by `ThurloeDao`

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala (hunk 5 above)
Comment (MatthewBemis) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala]: Addressed in 7d4194a6dc0d60f90e49aebff2cc6afb5d9e0702

Comment (tlangs) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala]: Thanks so much!

**Review by tlangs — APPROVED** [review]

**Pull request merged**